Added removing commands after X amount of seconds

thom-de-m opened the pull request:
So we had this discussion on the Shotbow Discord.
https://gyazo.com/bf82c7cf964043087f035a4332286091

And I thought I would add it with a merge request. If this is an unwanted feature just close / remove this pull request! It simply removes the command of the executor and matching messages after an X amount of seconds which can be defined in the config. Right now it's set to 60 seconds.

MistriHD commented:
"The StyleCI analysis has failed" but "StyleCI passed". What? xD

navarr commented:
If I were you I would suggest these as separate PRs (as they are different features).

StyleCI is also failing, and there are conflicts.

Either way, while in general you should suggest them as separate PRs - I don't think I'd approve either of them.

1) I don't really want more fun commands in ChatBot.
2) I like sending people over to mcstatus - we don't have to deal with handling any changes/upgrades/breakages related to the Minecraft status api

thom-de-m committed:
Added removing commands after X amount of seconds

thom-de-m committed:
Fixed mcstatus and playercount

Both commands were broken, that's now fixed

McJeffr commented:
I personally think that deleting all messages of ChatBot is a bad idea. This might not always be desired behaviour and some messages might need to be persistent (e.g. the PR of Mistri about warning and muting players). Instead, my suggestion would be to make a property for a command (just like the alias property, for instance) that declares whether or not the messages/command output need to be deleted after x amount of time, with not deleting them as the default behaviour.

For example, imagine someone asking a link to the forums. Someone responds with !forums, to answer that person. The message than gets deleted after x seconds. It then looks like noone has yet responded to this person, causing many more people to respond to this person who has already gotten his answer. In fact, Discord's chat is "async" in nature, meaning that someone might not always immediately see your response. ChatBot could delete the message before the person at whom the command is targetted has even seen it.

thom-de-m committed:
Added the shouldDeleteMessage property

McJeffr commented:
I am a bit confused why in the Playercount.js all messages are sent to `this.deleteMessage()` wheras adding `shouldDeleteMessage: true` to something like About.js makes it work without having to modify the internal logic of the command to use `this.deleteMessage()`. This seems inconsistent and a quick fix for dealing with the async problems that occur in Playercount.js (fetchServerlist). Some other solution to deal with the async call with a callback in Playercount.js might need to be found (e.g. turning the fetchServerlist into a promise and using async/await) so that all these additional parameters and functions are not needed as it creates inconsistency. Also, adding a timeOut property to the processMessage function seems a bit odd for this functionality as the message event of Discord is capable (and already is) of dealing with deleting messages, the command classes merely are dependents of this Command.js class and can access all of its functions through `this`.

EDIT: Essentially what I am trying to say is that in order for it to work automatically, the result of message.channel.send() needs to be returned, like in About.js. This message.channel.send() should also be returned in Playercount.js, which means that the async call to fetchServerlist needs to be addressed to somehow allow for the message to be returned as well. The async / await keywords in JS can help to achieve this. This would make everything consistent and removes the "quick fix".

Other points of feedback:
- A reference to `var self = this` is not needed when working with ES6 functions as `this` behaves differently from normal JS functions. See Command.js's initialize method.

thom-de-m commented:
> I am a bit confused why in the Playercount.js all messages are sent to this.deleteMessage() wheras adding shouldDeleteMessage: true to something like About.js makes it work without having to modify the internal logic of the command to use this.deleteMessage(). This seems inconsistent and a quick fix for dealing with the async problems that occur in Playercount.js (fetchServerlist). Some other solution to deal with the async call with a callback in Playercount.js might need to be found (e.g. turning the fetchServerlist into a promise and using async/await) so that all these additional parameters and functions are not needed as it creates inconsistency. Also, adding a timeOut property to the processMessage function seems a bit odd for this functionality as the message event of Discord is capable (and already is) of dealing with deleting messages, the command classes merely are dependents of this Command.js class and can access all of its functions through this.

I've created inside both `MCStatus.js` and `Playercount.js` now promises in combination with the async and await keyword. And with that removed the timeOut from the `processMessage` function.

Thanks for the feedback!

thom-de-m committed:
Cleaned up MCStatus & Playercount

Both now use promises instead of handling messages them selfs

McJeffr commented:
Ah, this looks much cleaner and easier to use. I have a few last points of feedback that involve some minor issues or optional improvements, nothing too major.

- Warning: When a user deletes their own message (which is targetted for removal), a warning is thrown in the console (though the bot does not crash). Might want to quickly catch this warning and discard it because it is no big deal.
- Code consistency: Though not everywhere, in most places ES6 syntax is used. This means using `const` and `let` instead of `var` as well as using ES6 functions.
- - In Playercount.js at line 63: removing `var self = this` and replacing all of `self` with `this` in the function, as well as turning the function in the promise into an ES6 function generally cleans up some of the code related to this consistency problem. (`return new Promise((resolve, reject) => { ... }`)
- - In MCStatus.js at line 28: Same story as with Playercount.js. You can see an arrow function on line 31 in this same file as well.
- - In Command.js at line 45: Use console.error instead of console.log. See Playercount.js at line 49.
- - In Command.js at line 19: Same as the previous `let self = this`, this isn't even needed as there is already an arrow function. You can see a mix of `this` and `self` in this arrow function which isn't really put there by you, though it is probably a good idea to convert all of this to `this` so we got that problem out of the way as well.
- - In Command.js at line 55: A space between `for` and `(let message in element) {`.
- Functionality: Feel free to add `shouldDeleteMessage: true` in the Playercount command as this is the main reason for this new functionality. The timeout of 60 seconds seems fine, it could even be reduced a bit to 30 seconds but we'll see how 60 seconds performs first.

thom-de-m committed:
Cleaned up some more code

Removed a lot of the 'self' instances,
and with that added a lot more 'this' references

Set 'shouldDeleteMessage' to true in the Playercount.js

JamsheedMistri commented:
You did a lot more than add a timeout, but I approve of the changes regardless — it's a nice cleanup to the messy code. Of course, since @McJeffr and I are not developers, we cannot approve this until @navarr reviews it. I will assign it to him now.

navarr commented on src/Command.js:
```diff
@@ -15,10 +16,9 @@ module.exports = BotModule.extend({
         this.discordClient.on('message', message => {
             let messageText = message.content;
-            if (messageText.substr(0, 1) !== self.commandPrefix) return;
+            if (messageText.substr(0, 1) !== this.commandPrefix) return;
```
Why did you change `self` to `this`?  Isn't `this` in this context dependant on the `.on` method?

McJeffr replied:
Arrow functions behave different than using functions with the function keyword. Arrow functions have a different context when using "this". See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions.

navarr commented on src/Command.js:
```diff
@@ -28,12 +28,35 @@ module.exports = BotModule.extend({
+                    return;
+                }
+                
+                message.delete(timeOut).catch(err => {/*do nothing*/});
```
It would probably be best to log such errors.  Why have you chosen not to?

McJeffr replied:
This error should only appear when a message is being deleted that was already deleted by a user. I am assuming it therefore was not necessary to log that error because it is not of any severity, the catch is mainly there to prevent stacktraces from popping up if this is the case.

navarr reviewed (changes requested):
Please ensure your files have newlines at the end of them.  I've also asked a couple questions which might not necessarily require changes - but that I would like the answers to

navarr commented:
@JamsheedMistri Are you able to verify the functionality works as expected?

navarr commented:
@McJeffr Thank you for answering my questions.  I'm actually fine merging this and moving forward if you or @JamsheedMistri can affirm it works

JamsheedMistri commented:
@navarr Can confirm it works.
